src/rp2_common/hardware_rtc/rtc.c

```diff
@@ -114,60 +139,88 @@ void rtc_enable_alarm(void) {
+    _callback();
+
+  // If it is a repeatable alarm, re enable the alarm.
+  if(_alarm_repeats) {
+        rtc_enable_alarm();
     }
 }
 
-static bool rtc_alarm_repeats(datetime_t *t) {
-    // If any value is set to -1 then we don't match on that value
-    // hence the alarm will eventually repeat
-    if (t->year  < 0) return true;
-    if (t->month < 0) return true;
-    if (t->day   < 0) return true;
-    if (t->dotw  < 0) return true;
-    if (t->hour  < 0) return true;
-    if (t->min   < 0) return true;
-    if (t->sec   < 0) return true;
-    return false;
+static repeat_type rtc_alarm_repeats(const datetime_t *t) {
+  // If any value is set to -1 then we don't match on that value
+  // hence the alarm will eventually repeat
+    if (t->year < 0 && t->month < 0 && t->day < 0 && t->dotw < 0
+        && t->hour < 0 && t->min < 0 && t->sec < 0) return CONTINUOUS_REPEAT_EVERY_SEC;
+
+    return (t->year < 0 || t->month < 0 || t->day < 0 || t->dotw < 0
+            || t->hour < 0 || t->min < 0 || t->sec < 0)
+        ? CONTINUOUS_REPEAT : NO_REPEAT;
 }
 
-void rtc_set_alarm(datetime_t *t, rtc_callback_t user_callback) {
-    rtc_disable_alarm();
+bool rtc_set_alarm(const datetime_t *t, rtc_callback_t user_callback) {
+    if (!rtc_running())
+      return false;
 
-    // Only add to setup if it isn't -1
-    rtc_hw->irq_setup_0 = ((t->year  < 0) ? 0 : (((uint)t->year)  << RTC_IRQ_SETUP_0_YEAR_LSB )) |
-                          ((t->month < 0) ? 0 : (((uint)t->month) << RTC_IRQ_SETUP_0_MONTH_LSB)) |
-                          ((t->day   < 0) ? 0 : (((uint)t->day)   << RTC_IRQ_SETUP_0_DAY_LSB  ));
-    rtc_hw->irq_setup_1 = ((t->dotw  < 0) ? 0 : (((uint)t->dotw)  << RTC_IRQ_SETUP_1_DOTW_LSB)) |
-                          ((t->hour  < 0) ? 0 : (((uint)t->hour)  << RTC_IRQ_SETUP_1_HOUR_LSB)) |
-                          ((t->min   < 0) ? 0 : (((uint)t->min)   << RTC_IRQ_SETUP_1_MIN_LSB )) |
-                          ((t->sec   < 0) ? 0 : (((uint)t->sec)   << RTC_IRQ_SETUP_1_SEC_LSB ));
+    rtc_disable_alarm();
 
-    // Set the match enable bits for things we care about
-    if (t->year  >= 0) hw_set_bits(&rtc_hw->irq_setup_0, RTC_IRQ_SETUP_0_YEAR_ENA_BITS);
-    if (t->month >= 0) hw_set_bits(&rtc_hw->irq_setup_0, RTC_IRQ_SETUP_0_MONTH_ENA_BITS);
-    if (t->day   >= 0) hw_set_bits(&rtc_hw->irq_setup_0, RTC_IRQ_SETUP_0_DAY_ENA_BITS);
-    if (t->dotw  >= 0) hw_set_bits(&rtc_hw->irq_setup_1, RTC_IRQ_SETUP_1_DOTW_ENA_BITS);
-    if (t->hour  >= 0) hw_set_bits(&rtc_hw->irq_setup_1, RTC_IRQ_SETUP_1_HOUR_ENA_BITS);
-    if (t->min   >= 0) hw_set_bits(&rtc_hw->irq_setup_1, RTC_IRQ_SETUP_1_MIN_ENA_BITS);
-    if (t->sec   >= 0) hw_set_bits(&rtc_hw->irq_setup_1, RTC_IRQ_SETUP_1_SEC_ENA_BITS);
+    uint32_t s0 = 0, s1 = 0;
 
     // Does it repeat? I.e. do we not match on any of the bits
     _alarm_repeats = rtc_alarm_repeats(t);
 
+    bool check_params = (is_valid_datetime(t) || _alarm_repeats != NO_REPEAT) && user_callback;
```

Comment:
@kilograham Do we definitely want to just `return false` here (and not set the alarm) if a `user_callback` wasn't supplied?

Reply:
ok this one is wrong - I fix it

Datasheet 4.8.3. says this event is also used to wakeup the pico

Reply:
Anyway I think the HAL should differ between a user event and a wakeup from dormant mode, because if it is in dormant mode, the user callback may not be executed as clk_sys may be 0

Reply:
ok no user callback in sleep because it's overwritten